NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/doc/clmdist.1
.\" Copyright (c) 2014 Stijn van Dongen
.TH "clm dist" 1 "16 May 2014" "clm dist 14-137" "USER COMMANDS "
.po 2m
.de ZI
.\" Zoem Indent/Itemize macro I.
.br
'in +\\$1
.nr xa 0
.nr xa -\\$1
.nr xb \\$1
.nr xb -\\w'\\$2'
\h'|\\n(xau'\\$2\h'\\n(xbu'\\
..
.de ZJ
.br
.\" Zoem Indent/Itemize macro II.
'in +\\$1
'in +\\$2
.nr xa 0
.nr xa -\\$2
.nr xa -\\w'\\$3'
.nr xb \\$2
\h'|\\n(xau'\\$3\h'\\n(xbu'\\
..
.if n .ll -2m
.am SH
.ie n .in 4m
.el .in 8m
..
.SH NAME
clm dist \- compute the distance between two or more partitions (clusterings)\&.

The distance that is computed can be any of
\fIsplit/join distance\fP, \fIvariance of information\fP,
or \fIMirkin metric\fP\&.

clmdist is not in actual fact a program\&. This manual
page documents the behaviour and options of the clm program when
invoked in mode \fIdist\fP\&. The options \fB-h\fP, \fB--apropos\fP,
\fB--version\fP, \fB-set\fP, \fB--nop\fP are accessible
in all \fBclm\fP modes\&. They are described
in the \fBclm\fP manual page\&.
.SH SYNOPSIS

\fBclm dist\fP [options] <file name> <file name>+

\fBclm dist\fP
\fB[-mode\fP <sj|vi|mk|sc> (\fIdistance type\fP)\fB]\fP
\fB[-o\fP fname (\fIoutput file\fP)\fB]\fP
\fB[--chain\fP (\fIonly compare consecutive clusterings\fP)\fB]\fP
\fB[--one-to-many\fP (\fIcompare first clustering to all others\fP)\fB]\fP
\fB[--sort\fP (\fIsort clusterings based on coarseness\fP)\fB]\fP
\fB[--index\fP (\fIoutput Rand, adjusted Rand and Jaccard indices\fP)\fB]\fP
\fB[-digits\fP k (\fIoutput decimals\fP)\fB]\fP
\fB[-h\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--apropos\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--version\fP (\fIprint version, exit\fP)\fB]\fP
<file name> <file name>+
.SH DESCRIPTION

\fBclm dist\fP computes distances between clusterings\&. It can compute the
\fIsplit/join distance\fP (described below), the \fIvariance of information
measure\fP, and the \fIMirkin metric\fP\&. By default it computes the chosen distance
for all pairs of distances in the clusterings provided\&. Clusterings must be in
the mcl matrix format (cf\&. \fBmcxio(5)\fP), and are supplied on the command
line as the names of the files in which they are stored\&.
It is possible to compare only consecutive clusterings by using
the \fB--chain\fP option\&.

Currently, \fBclm dist\fP cannot compute different distance types simultaneously\&.

The output is linewise, each line giving information about
the distance between a pair of clusterings\&. A line has the
following format:

.di ZV
.in 0
.nf \fC
d  d1  d2  N  v  name1  name2  [v50,v75,v90,v95,v99]
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

where \fCdT\fP is the distance between the two clusterings, \fCd1\fP is the
distance from the first clustering to the greatest common subclustering
(alternatively called GCS, intersection, or meet) of the two clusterings,
\fCd2\fP is similarly the distance from the second clustering to the GCS,
\fCN\fP is the number of nodes in the set over which the clusterings are
defined, \fCname1\fP is the name of the file containing the first clustering,
\fCname2\fP is the name of the file containing the second clustering, and
\fCvXX\fP is the number of \fIvolatile nodes\fP at stringency factor \fC0\&.XX\fP
(i\&.e\&. 0\&.5 for \fCv50\fP)\&. Refer to \fBclm\ \&vol\fP for a definition of
\fIvolatile node\fP\&.
.SH OPTIONS

.ZI 2m "\fB-mode\fP <sj|vi|mk> (\fIdistance type\fP)"
\&
.br
Use \fBsj\fP for the \fIsplit/join distance\fP (described below), \fBvi\fP for
the \fIvariance of information measure\fP and \fBmk\fP for the \fIMirkin metric\fP\&.
.in -2m

.ZI 2m "\fB--chain\fP (\fIonly compare consecutive clusterings\fP)"
\&
.br
This option can be used if you know that the clusterings are nested
clusterings (or appoximately so) and ordered from coarse to fine-grained
or vice versa\&. An example of this is the set of clusterings resulting
from applying \fBmcl\fP with a range of inflation parameters\&.
.in -2m

.ZI 2m "\fB--one-to-many\fP (\fIcompare first clustering to all others\fP)"
\&
.br
Use this option for example to compare a gold standard classification
to a collection of clusterings\&.
Bear in mind that sub-clustering and super-clustering are also
ways for a clustering to be compatible with a gold standard\&.
This means that the simple numerical criterion of distance between
clusters (by whatever method) is only partially informative\&.
For the Mirkin, variation of information and split/join metrics
it pays to take into account the constituent distances \fCd1\fP
and \fCd2\fP (see above)\&. Assuming that the first clustering
given as argument represents a gold standard, a small value
for \fCd1\fP implies that the second clustering is (nearly) a superclustering,
and similarly a small value for \fCd2\fP implies that it is (nearly)
a subclustering\&.
.in -2m

.ZI 2m "\fB--sort\fP (\fIsort clusterings based on coarseness\fP)"
\&
.br
This option can be useful in conjunction with the \fB--chain\fP
option, in case the list of clusterings supplied is not necessarily
ordered by granularity\&.
.in -2m

.ZI 2m "\fB--index\fP (\fIoutput Rand, adjusted Rand and Jaccard indices\fP)"
\&
.br
As described\&.
.in -2m

.ZI 2m "\fB-o\fP fname (\fIoutput file\fP)"
\&
.br
.in -2m

.ZI 2m "\fB-digits\fP k (\fIoutput decimals\fP)"
\&
.br
The number of decimals printed when using the variance of information measure\&.
.in -2m
.SH SPLIT/JOIN DISTANCE

For each pair of clusterings \fBC1\fP, \fBC2\fP, two numbers are given,
say \fBd1\fP and \fBd2\fP\&. Then \fBd1\fP + \fBd2\fP equals the number
of nodes that have to be exchanged in order to transform any of the two
clusterings into the other, and you can think of (\fBd1\fP+\fBd2\fP)/\fB2N\fP
as the percentage that the two clusterings differ\&. The split/join
distance has a linearity property with respect to the meet of \fBC1\fP and
\fBC2\fP, see below\&.

The split/join distance \fBsjd\fP is very handy in computing the consistency of
two or more clusterings of the same graph, or comparing clusterings made
with different resource (but otherwise identical) parameters\&. The latter is
for finding out whether you can settle for cheaper mcl settings, or whether
you need to switch to more expensive settings\&. The former is for finding out
whether clusterings are identical, conflicting, or whether one is (almost) a
subclustering of the other - mostly for comparing a set of clusterings of
different granularity, made by letting the mcl parameter \fB-I\fP vary\&.
The \fBEXAMPLES\fP section contains examples of all these \fBclm dist\fP uses,
and the use of \fBclm info\fP and \fBclm meet\fP is also discussed there\&.

\fBsjd\fP is a metric distance on the space of partitions of
a set of a given fixed cardinality\&. It has the following linearity
property\&. Let \fBP1\fP and \fBP2\fP be partitions, then

\fBsjd\fP(\fBP1\fP, \fBP2\fP) = \fBsjd\fP(\fBP1\fP, \fBD\fP) + \fBsjd\fP(\fBP2\fP, \fBD\fP)

where \fBD\fP (for Dutch Doorsnede)
is the intersection of \fBP1\fP and \fBP2\fP, i\&.e\&. the unique clustering
that is both a subclustering of \fBP1\fP and \fBP2\fP \fIand\fP a superclustering of
all other subclusterings of \fBP1\fP and \fBP2\fP\&. Sloppily worded, \fBD\fP is the largest
subclustering of both \fBP1\fP and \fBP2\fP\&. See the \fBREFERENCES\fP section for
a pointer to the technical report in which \fBsjd\fP was first defined (and in
which the non-trivial triangle inequality is proven)\&.

Because it is useful to know whether one partition (or clustering)
is almost a subclustering of the other, \fBclm dist\fP returns the
two constituents \fBsjd\fP(\fBP1\fP,\fBD\fP) and \fBsjd\fP(\fBP2\fP,\fBD\fP)\&.

Let \fBP1\fP and \fBP2\fP be two clusterings of a graph of cardinality \fBN\fP,
and suppose \fBclm dist\fP returns the integers \fBd1\fP and \fBd2\fP\&. You can think of
\fB100 * (d1 + d2) / N\fP as the percentage that \fBP1\fP and \fBP2\fP differ\&.
This interpretation is in fact slightly conservative\&.
The numerator is the number of nodes that need to be exchanged in order to
transform one into the other\&. This number may grow as large as
\fB2*N - 2*sqrt(N)\fP, so it would be justified to take 50 as a scaling
factor rather than 100\&.

For example, if \fBA\fP and \fBB\fP are both clusterings of a graph
on a set of 9058 nodes and \fBclm dist\fP returns [38, 2096], this conveys
that \fBA\fP is almost a subclustering of \fBB\fP (by splitting 38 nodes
in \fBA\fP we obtain a clustering \fBD\fP that is a subclustering of \fBB\fP),
and that \fBB\fP is much less granular than \fBA\fP\&. The latter is
because we can obtain \fBB\fP from \fBD\fP by \fIjoining\fP 2096 nodes
in some way\&.
.SH EXAMPLES

The following is an example of several mcl validation tools
applied to a set of clusterings on a protein graph of 9058 nodes\&.
In the first experiment, six
different clusterings were generated for different values of the inflation
parameter, which was respectively set to 1\&.2, 1\&.6, 2\&.0, 2\&.4, 2\&.8, and 3\&.2\&.
It should be noted that protein graphs seem somewhat special in that an
inflation parameter setting as low as 1\&.2 still produces a very acceptable
clustering\&. The six clusterings are scrutinized using \fBclm dist\fP,
\fBclm info\fP, and \fBclm meet\fP\&.
In the second experiment, four different clusterings were generated
with identical flow (i\&.e\&. inflation) parameter, but
with different resource parameters\&. \fBclm dist\fP is used to choose
a sufficient resource level\&.

High \fB-P/-S/-R\fP values make \fBmcl\fP more accurate but also
more time and memory consuming\&. Run \fBmcl\fP with different settings for these
parameters, holding other parameters fixed\&. If the expensive and supposedly
more accurate clusterings are very similar to the clusterings resulting from
cheaper settings, the cheaper setting is sufficient\&. If the distances
between cheaper clusterings and more expensive clusterings are large, this
is an indication that you need the expensive settings\&. In that case, you may
want to increase the \fB-P/-S/-R\fP parameters (or simply the
\fB-scheme\fP parameter) until associated
clusterings at nearby resource levels are very similar\&.

In this particular example, the validation tools do not reveal that one
clustering in particular can be chosen as \&'best\&', because all clusterings
seem at least acceptable\&. They do aid however in showing the relative
merits of each clusterings\&. The most important issue in this respect is
cluster granularity\&. The table below shows the output of \fBclm info\fP\&.

.di ZV
.in 0
.nf \fC

     Efficiency  Mass frac  Area frac  Cl weight  Mx link weight
1\&.2   0\&.42364     0\&.98690    0\&.02616    52\&.06002    50\&.82800
1\&.6   0\&.58297     0\&.95441    0\&.01353    55\&.40282    50\&.82800
2\&.0   0\&.63279     0\&.92386    0\&.01171    58\&.09409    50\&.82800
2\&.4   0\&.65532     0\&.90702    0\&.01091    59\&.58283    50\&.82800
2\&.8   0\&.66854     0\&.84954    0\&.00940    63\&.19183    50\&.82800
3\&.2   0\&.67674     0\&.82275    0\&.00845    66\&.10831    50\&.82800
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

This data shows that there is exceptionally strong cluster structure present
in the input graph\&. The 1\&.2 clustering captures almost all edge mass using
only 2\&.5 percent of \&'area\&'\&. The 3\&.2 clustering still captures 82 percent of
the mass using less than 1 percent of area\&. We continue with looking at the
mutual consistency of the six clusterings\&. Below is a table that shows all
pairwise distances between the clusterings\&.

.di ZV
.in 0
.nf \fC

    |   1\&.6  |   2\&.0  |   2\&.4  |   2\&.8  |   3\&.2  |   3\&.6
-----------------------------------------------------------\&.
1\&.2 |2096,38 |2728,41 |3045,48 |3404,45 |3621,43 |3800, 42 |
-----------------------------------------------------------|
1\&.6 |        | 797,72 |1204,76 |1638,78 |1919,70 |2167, 69 |
-----------------------------------------------------------|
2\&.0 |        |        | 477,68 | 936,78 |1235,85 |1504, 88 |
-----------------------------------------------------------|
2\&.4 |        |        |        | 498,64 | 836,91 |1124,103 |
-----------------------------------------------------------|
2\&.8 |        |        |        |        | 384,95 | 688,119 |
-----------------------------------------------------------|
3\&.2 |        |        |        |        |        | 350,110 |
-----------------------------------------------------------\&.
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

The table shows that the different clusterings are pretty consistent with
each other, because for two different clusterings it is generally true that
one is almost a subclustering of the other\&. The interpretation for the
distance between the 1\&.6 and the 3\&.2 clustering for example, is that by
rearranging 43 nodes in the 3\&.2 clustering, we obtain a subclustering of the
1\&.6 clustering\&. The table shows that for any pair of clusterings, at most
119 entries need to be rearranged in order to make one a subclustering of
the other\&.

The overall consistency becomes all the more clear by looking at the meet of
all the clusterings:

.di ZV
.in 0
.nf \fC

clm meet -o meet out12 out16 out20 out24 out28 out32
clm dist meet out12 out16 out20 out24 out28 out32
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

results in the following distances between the respective clusterings
and their meet\&.

.di ZV
.in 0
.nf \fC

    |   1\&.2  |    1\&.6 |  2\&.0   |   2\&.4  |  2\&.8   |  3\&.2    |  
-------------- --------------------------------------------\&.
meet|  0,3663|  0,1972| 0,1321 |  0,958 | 0,559  | 0,283   |
-------------- --------------------------------------------\&.
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

This shows that by rearranging only 283 nodes in the 3\&.2 clustering,
one obtains a subclustering of all other clusterings\&.

In the last experiment, \fBmcl\fP was run with inflation parameter 1\&.4,
for each of the four different preset pruning schemes \fCk=1,2,3,4\fP\&.
The \fBclm dist\fP distances between the different clusterings
are shown below\&.

.di ZV
.in 0
.nf \fC

    |  k=2   |   k=3  |   k=4  |
-------------------------------\&.
k=1 |  17,17 |  16,16 |  16,16 |
-------------------------------\&.
k=2 |        |   3,3  |   5,5  |
-------------------------------\&.
k=3 |        |        |   4,4  |
-------------------------------\&.
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

This example is a little boring in that the cheapest scheme seems adequate\&.
If anything, the gaps between the \fCk=1\fP scheme and the rest are a little
larger than the three gaps between the \fCk=2\fP, \fCk=3\fP, and \fCk=4\fP
clusterings\&. Had all distances been much larger, then such an observation
would be reason to choose the \fCk=2\fP setting\&.


NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/src/shcl/clmdist.c
/*   (C) Copyright 2001, 2002, 2003, 2004, 2005 Stijn van Dongen
 *   (C) Copyright 2006, 2007, 2008, 2009, 2010 Stijn van Dongen
 *   (C) Copyright 2011, 2012, 2013  Stijn van Dongen
 *
 * This file is part of MCL.  You can redistribute and/or modify MCL under the
 * terms of the GNU General Public License; either version 3 of the License or
 * (at your option) any later version.  You should have received a copy of the
 * GPL along with MCL, in the file COPYING.
*/

/* TODO
 *    rand index, randc (Hubert et ??)

   (n:2) + 2 Sum{i=1->c1} Sum{j=1->c2} (n_ij:2) - [ Sum{i=1->c1} (n_i+:2) + Sum{j=1->c2} (n_+j:2) ]
                                 -------------------------
                                       n (n-1) : 2
     a + d
   ---------
     (n/2)

   a: same in C1 and same in C2
   d: different in C1 and different in C2.


            Sum{i=1->c1} Sum{j=1->c2} (n_ij/2) - [1/(n/2)] Sum{i=1->c1} (n_i/2) * Sum{j=1->c2} (n_j/2)
                                 -------------------------
 [1/2] *  ( Sum{i=1->c1} (n_i/2) + Sum{j=1->c2} (n_j/2) ) - [1/(n/2)] Sum{i=1->c1} (n_i/2) * Sum{j=1->c2} (n_j/2)

*/

#include <string.h>
#include <stdio.h>
#include <limits.h>

#include "clm.h"
#include "report.h"
#include "clmdist.h"

#include "impala/matrix.h"
#include "impala/io.h"
#include "impala/iface.h"
#include "impala/ivp.h"
#include "impala/app.h"
#include "impala/io.h"

#include "clew/clm.h"
#include "clew/cat.h"

#include "util/types.h"
#include "util/err.h"
#include "util/opt.h"



static double mclv_choose_sum
(  const mclv* vec
)  
   {  mclIvp* vecivps = vec->ivps
   ;  dim     vecsize = vec->n_ivps
   ;  double  sum  = 0.0

   ;  while (vecsize-- > 0)
      {  double val = (vecivps++)->val
      ;  double delta = val * (val - 1.0) * 0.5
      ;  if (delta > 0)
         sum += delta
   ;  }
      return sum
;  }


static double mclx_choose_sum
(  const mclx* mx
)
   {  dim i
   ;  double sum = 0.0
   ;  for (i=0;i<N_COLS(mx);i++)
      sum += mclv_choose_sum(mx->cols+i)
   ;  return sum
;  }


enum
{  DIST_SPLITJOIN 
,  DIST_VARINF
,  DIST_MIRKIN
,  INDEX
}  ;


/*
 * clmdist will enstrict the clusterings, and possibly project them onto
 * the meet of the domains if necessary.
*/


static const char* me = "clm dist";


enum
{  DIST_OPT_OUTPUT = CLM_DISP_UNUSED
,  DIST_OPT_MODE
,  DIST_OPT_INDEX
,  DIST_OPT_CHAIN
,  DIST_OPT_WHEEL
,  DIST_OPT_SORT
,  DIST_OPT_NORMALISE
,  DIST_OPT_MCI
,  DIST_OPT_DIGITS
,  DIST_OPT_FACTOR
}  ;


enum
{  VOL_OPT_OUTPUT = CLM_DISP_UNUSED
,  VOL_OPT_FACTOR
}  ;

static mcxOptAnchor volOptions[] =
{  {  "-o"
   ,  MCX_OPT_HASARG
   ,  VOL_OPT_OUTPUT
   ,  "<fname>"
   ,  "output file name"
   }
,  {  "-fraction"
   ,  MCX_OPT_HASARG
   ,  VOL_OPT_FACTOR
   ,  "<num>"
   ,  "stringency factor: require |meet| >= <num> * |cls-size| (default 0.5)"
   }
,  {  NULL ,  0 ,  0 ,  NULL, NULL}
}  ;


static mcxOptAnchor distOptions[] =
{  {  "-o"
   ,  MCX_OPT_HASARG
   ,  DIST_OPT_OUTPUT
   ,  "<fname>"
   ,  "output file name"
   }
,  {  "-mode"
   ,  MCX_OPT_HASARG
   ,  DIST_OPT_MODE
   ,  "<mode>"
   ,  "one of sj|vi|mirkin"
   }
,  {  "--mci"
   ,  MCX_OPT_DEFAULT | MCX_OPT_HIDDEN
   ,  DIST_OPT_MCI
   ,  NULL
   ,  "output all against all matrix of distances"
   }
,  {  "--index"
   ,  MCX_OPT_DEFAULT
   ,  DIST_OPT_INDEX
   ,  NULL
   ,  "output Rand, corrected Rand and Jaccard index"
   }
,  {  "-digits"
   ,  MCX_OPT_HASARG
   ,  DIST_OPT_DIGITS
   ,  "<num>"
   ,  "number of trailing digits for floats"
   }
,  {  "--chain"
   ,  MCX_OPT_DEFAULT
   ,  DIST_OPT_CHAIN
   ,  NULL
   ,  "only compare consecutive clusterings"
   }
,  {  "--one-to-many"
   ,  MCX_OPT_DEFAULT
   ,  DIST_OPT_WHEEL
   ,  NULL
   ,  "compare first clustering against all the rest"
   }
,  {  "--sort"
   ,  MCX_OPT_DEFAULT
   ,  DIST_OPT_SORT
   ,  NULL
   ,  "sort from coarse to fine-grained before computing distances"
   }
,  {  "--normalise"
   ,  MCX_OPT_DEFAULT | MCX_OPT_HIDDEN    /* not implemented */
   ,  DIST_OPT_NORMALISE
   ,  NULL
   ,  "NOT IMPLEMENTED normalise criterion (applicable for --vi and default split/join)"
   }
,  {  "-fraction"
   ,  MCX_OPT_HASARG | MCX_OPT_HIDDEN
   ,  DIST_OPT_FACTOR
   ,  "<num>"
   ,  "stringency factor: require |meet| >= <num> * |cls-size| (default 0.5)"
   }
,  {  NULL ,  0 ,  0 ,  NULL, NULL}
}  ;


static mcxIO*  xfout    =  (void*) -1;
static int digits       =  -1;
static int mode_g       =  -1;
static mcxbool i_am_vol =  FALSE;   /* node faithfulness */
static double  nff_fac  =  FLT_MAX;
static mcxbool consecutive_g = -1;
static mcxbool mci_g = -1;
static mcxbool wheel_g = -1;
static mcxbool split_g = -1;
static mcxbool sort_g = -1;


static mcxstatus distInit
(  void
)
   {  xfout =  mcxIOnew("-", "w")
   ;  mode_g = 0
   ;  digits = 2
   ;  nff_fac     =  0.5
   ;  consecutive_g = FALSE
   ;  mci_g = FALSE
   ;  wheel_g = FALSE
   ;  sort_g = FALSE
   ;  split_g = FALSE
   ;  return STATUS_OK
;  }


static mcxstatus volInit
(  void
)
   {  xfout       =  mcxIOnew("-", "w")
   ;  i_am_vol    =  TRUE
   ;  nff_fac     =  0.5
   ;  consecutive_g = FALSE
   ;  mci_g = FALSE
   ;  wheel_g = FALSE
   ;  sort_g = FALSE
   ;  return STATUS_OK
;  }


static mcxstatus volArgHandle
(  int optid
,  const char* val
)
   {  switch(optid)
      {  case VOL_OPT_FACTOR
      :  nff_fac = atof(val)
      ;  break
      ;

         case VOL_OPT_OUTPUT
      :  mcxIOnewName(xfout, val)
      ;  break
      ;

         default
      :  return STATUS_FAIL
      ;
      }
      return STATUS_OK
;  }


static mcxstatus distArgHandle
(  int optid
,  const char* val
)
   {  switch(optid)
      {  case DIST_OPT_OUTPUT
      :  mcxIOnewName(xfout, val)
      ;  break
      ;

         case DIST_OPT_DIGITS
      :  digits = atoi(val)
      ;  break
      ;

         case DIST_OPT_WHEEL
      :  wheel_g = TRUE
      ;  break
      ;

         case DIST_OPT_CHAIN
      :  consecutive_g = TRUE
      ;  break
      ;

         case DIST_OPT_MCI
      :  mci_g = TRUE
      ;  break
      ;

         case DIST_OPT_SORT
      :  sort_g = TRUE
      ;  break
      ;

         case DIST_OPT_FACTOR
      :  nff_fac = atof(val)
      ;  break
      ;

         case DIST_OPT_MODE
      :     if (!strcmp(val, "sj"))
            mode_g = DIST_SPLITJOIN
         ;  else if (!strcmp(val, "vi"))
            mode_g = DIST_VARINF
         ;  else if
            (  !strcmp(val, "ehd")
            || !strcmp(val, "mk")
            || !strcmp(val, "mirkin")
            )
            mode_g = DIST_MIRKIN
         ;  else
            mcxDie(1, me, "unknown mode <%s>", val)
      ;  break
      ;

         case DIST_OPT_INDEX
      :  mode_g = INDEX
      ;  break
      ;

         default
      :  return STATUS_FAIL
      ;
      }
      return STATUS_OK
;  }


static mcxstatus distMain
(  int                  argc
,  const char*          argv[]
)
   {  int               i
   ;  int a             =  0
   ;  mclx* nff_scores  =  NULL
   ;  mcxIO* xfin       =  mcxIOnew("-", "r")
   ;  mcxbits bits      =  MCLX_PRODUCE_PARTITION | MCLX_REQUIRE_DOMSTACK

   ;  mclxCat st
   ;  mclxCat st2

   ;  mclxCat *stptr = &st, *stptr1 = &st, *stptr2 = &st

   ;  mcxIO* xfdebug = mcxIOnew("-", "w")
   ;  mcxIOopen(xfdebug, EXIT_ON_FAIL)

   ;  mclxCatInit(&st)
   ;  mclxCatInit(&st2)

   ;  if (i_am_vol)
      me = "clm vol"
      
   ;  if (!mode_g)
      mode_g = DIST_SPLITJOIN

   ;  if (mode_g & DIST_MIRKIN || mode_g & DIST_SPLITJOIN)
      digits = 0

   ;  mcxIOopen(xfout, EXIT_ON_FAIL)

   ;  for (a=0;a<argc;a++)
      {  mcxstatus status
      ;  if (!strcmp(argv[a], "--"))
         {  stptr = &st2
         ;  split_g = TRUE
         ;  stptr2 = &st2
         ;  continue
      ;  }
         mcxIOnewName(xfin, argv[a])
      ;  status = mclxCatRead(xfin, stptr, 0, NULL, NULL, bits)
      ;  mcxIOclose(xfin)
      ;  if (status)
         break
;if (0)
         mclxWrite(stptr->level[stptr->n_level-1].mx, xfdebug, MCLXIO_VALUE_GETENV, RETURN_ON_FAIL)
   ;  }

      if (!a || a != argc)
      mcxDie(1, me, "failed to read one or more cluster files")

   ;  mcxIOfree(&xfdebug)

   ;  if (sort_g)
      mclxCatSortCoarseFirst(&st)

   ;  if (i_am_vol && st.n_level)
      nff_scores
      =  mclxCartesian
         (  mclvCanonical(NULL, 1, 1.0)
         ,  mclvClone(st.level[0].mx->dom_rows)
         ,  1.0
         )

   ;  if (mci_g)

   ;  for (i=0;i<stptr1->n_level;i++)
      {  mclx* c1       =  stptr1->level[i].mx
      ;  int j, jstart  =  split_g ? 0 : i+ 1
      ;  for (j=jstart; j<stptr2->n_level;j++)
         {  mclx* c2  =  stptr2->level[j].mx
         ;  mclx* meet12, *meet21
         ;  double dist1d, dist2d
         ;  dim dist1i, dist2i
         ;  dim n_volatile = 0
         ;  dim nff[5] = { 0, 0, 0, 0, 0 }

         ;  meet12 =  clmContingency(c1, c2)
         ;  meet21 =  mclxTranspose(meet12)

         ;  {  dim k
            ;  for (k=0;k<N_COLS(meet12);k++)
               {  dim l
               ;  mclv* ct = meet12->cols+k
               ;  mclv* c1mem = c1->cols+k
               ;  mclv* c2mem = NULL
               ;  for (l=0;l<ct->n_ivps;l++)
                  {  ofs c2id = ct->ivps[l].idx
                  ;  double meet_sz = ct->ivps[l].val
                  ;  c2mem = mclxGetVector(c2, c2id, EXIT_ON_FAIL, c2mem)
                  ;  if
                     (  meet_sz < nff_fac * c1mem->n_ivps
                     && meet_sz < nff_fac * c2mem->n_ivps
                     )
                     {  mclv* meet = mcldMeet(c1mem, c2mem, NULL)
                     ;  mclp* tivp = NULL
                     ;  dim m
                     ;  if (i_am_vol)
                        for (m=0;m<meet->n_ivps;m++)
                        {  tivp = mclvGetIvp(nff_scores->cols+0, meet->ivps[m].idx, tivp)
                        ;  tivp->val += 1 / (double) meet->n_ivps
                     ;  }
                        n_volatile += meet->n_ivps
                     ;  if (meet->n_ivps != meet_sz)
                        mcxErr(me, "meet size difference")
                     ;  mclvFree(&meet)
                  ;  }
                     if
                     (  meet_sz < 0.5 * c1mem->n_ivps
                     && meet_sz < 0.5 * c2mem->n_ivps
                     )
                     nff[0] += meet_sz
                  ;  if
                     (  meet_sz < 0.75 * c1mem->n_ivps
                     && meet_sz < 0.75 * c2mem->n_ivps
                     )
                     nff[1] += meet_sz
                  ;  if
                     (  meet_sz < 0.90 * c1mem->n_ivps
                     && meet_sz < 0.90 * c2mem->n_ivps
                     )
                     nff[2] += meet_sz
                  ;  if
                     (  meet_sz < 0.95 * c1mem->n_ivps
                     && meet_sz < 0.95 * c2mem->n_ivps
                     )
                     nff[3] += meet_sz
                  ;  if
                     (  meet_sz < 0.99999 * c1mem->n_ivps
                     && meet_sz < 0.99999 * c2mem->n_ivps
                     )
                     nff[4] += meet_sz
               ;  }
               }
               if (i_am_vol)  /* hacked; refactor later */
               continue
         ;  }

            if (mode_g == DIST_SPLITJOIN)
               clmSJDistance(c1, c2, meet12, meet21, &dist1i, &dist2i)
            ,  fprintf
               (  xfout->fp
               ,  "d=%lu\td1=%lu\td2=%lu\tnn=%ld\tc1=%ld\tc2=%ld\tv=%ld\tn1=%s\tn2=%s\tvol=[%ld,%ld,%ld,%ld,%ld]"
               ,  (ulong) (dist1i + dist2i)
               ,  (ulong) dist1i
               ,  (ulong) dist2i
               ,  (long) N_ROWS(c1)
               ,  (long) N_COLS(c1)
               ,  (long) N_COLS(c2)
               ,  (long) n_volatile
               ,  stptr1->level[i].fname->str
               ,  stptr2->level[j].fname->str
               ,  (long) nff[0]
               ,  (long) nff[1]
               ,  (long) nff[2]
               ,  (long) nff[3]
               ,  (long) nff[4]
               )

         ;  else if (mode_g == DIST_VARINF)
               clmVIDistance(c1, c2, meet12, &dist1d, &dist2d)
            ,  fprintf
               (  xfout->fp
               ,  "d=%.3f\td1=%.3f\td2=%.3f\tnn=%ld\tc1=%ld\tc2=%ld\tn1=%s\tn2=%s"
               ,  dist1d + dist2d
               ,  dist1d
               ,  dist2d
               ,  (long) N_ROWS(c1)
               ,  (long) N_COLS(c1)
               ,  (long) N_COLS(c2)
               ,  stptr1->level[i].fname->str
               ,  stptr2->level[j].fname->str
               )

         ;  else if (mode_g == DIST_MIRKIN)
               clmMKDistance(c1, c2, meet12, &dist1i, &dist2i)
            ,  fprintf
               (  xfout->fp
               ,  "d=%ld\td1=%ld\td2=%ld\tnn=%ld\tc1=%ld\tc2=%ld\tn1=%s\tn2=%s"
               ,  dist1i + dist2i
               ,  dist1i
               ,  dist2i
               ,  (long) N_ROWS(c1)
               ,  (long) N_COLS(c1)
               ,  (long) N_COLS(c2)
               ,  stptr1->level[i].fname->str
               ,  stptr2->level[j].fname->str
               )

         ;  else if (mode_g == INDEX)
            {  mclv* sizes_a  =  mclxColSizes(c1, MCL_VECTOR_SPARSE)
            ;  mclv* sizes_b  =  mclxColSizes(c2, MCL_VECTOR_SPARSE)
            ;  double soba    =  mclv_choose_sum(sizes_a)   /* sum of binomial */
            ;  double sobb    =  mclv_choose_sum(sizes_b)   /* sum of binomial */
            ;  double sobm    =  mclx_choose_sum(meet12)
            ;  double chsn_   =  (N_ROWS(c1) * 0.5 * (N_ROWS(c1) - 1.0))
            ;  double chsn    =  chsn_ ? chsn_ : 0.001   /* crazy, yes */

            ;  double rand    =  (chsn -  soba -  sobb +  2.0 * sobm) / chsn
            ;  double jaccard =  sobm / (soba + sobb - sobm)
            ;  double randc   =     (sobm - (soba * sobb) / chsn)
                                 /  ( 0.5 * (soba + sobb) - (soba * sobb) / chsn)
            ;  fprintf
               (  xfout->fp, "rand=%.5f jaccard=%.5f arand=%.5f n1=%s n2=%s"
               ,  rand, jaccard, randc
               ,  stptr1->level[i].fname->str ,  stptr2->level[j].fname->str
               )
         ;  }

            {  const char* e = getenv("CLMDIST_TAG")
            ;  if (e)
               fprintf(xfout->fp, "\ttag=%s", e)
            ;  fputc('\n', xfout->fp)
         ;  }

            mclxFree(&meet12)
         ;  mclxFree(&meet21)
         ;  if (consecutive_g)
            break
      ;  }
         if (wheel_g)
         break
   ;  }

      if (i_am_vol)
      {  mclxaWrite(nff_scores, xfout, 4, RETURN_ON_FAIL)
   ;  }
      return STATUS_OK
;  }


mcxDispHook* mcxDispHookDist
(  void
)
   {  static mcxDispHook distEntry
   =  {  "dist"
      ,  "dist [options] <cl file>+"
      ,  distOptions
      ,  sizeof(distOptions)/sizeof(mcxOptAnchor) - 1
      ,  distArgHandle
      ,  distInit
      ,  distMain
      ,  0
      ,  -1
      ,  MCX_DISP_MANUAL
      }
   ;  return &distEntry
;  }


mcxDispHook* mcxDispHookVol
(  void
)
   {  static mcxDispHook volEntry
   =  {  "vol"
      ,  "vol [options] <cl file>+"
      ,  volOptions
      ,  sizeof(volOptions)/sizeof(mcxOptAnchor) - 1
      ,  volArgHandle
      ,  volInit
      ,  distMain
      ,  1
      ,  -1
      ,  MCX_DISP_MANUAL
      }
   ;  return &volEntry
;  }


